From 83da18051d1bacb4d6b549b95f750e6d7fcaf647 Mon Sep 17 00:00:00 2001
From: "Vivek.A" <Vivek_Arumugam@comcast.com>
Date: Tue, 7 Feb 2023 08:28:01 +0000
Subject: [PATCH] comcast - RDK-57915 Analyze higher CPU usage of Web & Network
 processes

This change helps to recude WebProcess and NetworkProcess CPU usage,
expecially for Prime Video long video playbac

Change-Id: Id0ee2e2ab663b7c8d35a366c8883be34c688bfcb
---
 Source/WebCore/xml/XMLHttpRequest.cpp                  | 5 +++++
 Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp | 7 ++++++-
 2 files changed, 11 insertions(+), 1 deletion(-)

diff --git a/Source/WebCore/xml/XMLHttpRequest.cpp b/Source/WebCore/xml/XMLHttpRequest.cpp
index a7afbcee7a4e..d0483f2be3ff 100644
--- a/Source/WebCore/xml/XMLHttpRequest.cpp
+++ b/Source/WebCore/xml/XMLHttpRequest.cpp
@@ -1015,6 +1015,11 @@ void XMLHttpRequest::didSendData(unsigned long long bytesSent, unsigned long lon
 void XMLHttpRequest::didReceiveResponse(ScriptExecutionContextIdentifier, ResourceLoaderIdentifier, const ResourceResponse& response)
 {
     m_response = response;
+
+    if (response.tainting() != ResourceResponseBase::Tainting::Opaque &&
+            response.tainting() != ResourceResponseBase::Tainting::Opaqueredirect &&
+            readyState() < HEADERS_RECEIVED)
+       changeState(HEADERS_RECEIVED);
 }
 
 static inline bool shouldDecodeResponse(XMLHttpRequest::ResponseType type)
diff --git a/Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp b/Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp
index 157912192f22..f5062cad9bd5 100644
--- a/Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp
+++ b/Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp
@@ -254,7 +254,12 @@ void WebLoaderStrategy::scheduleLoad(ResourceLoader& resourceLoader, CachedResou
     }
 
     WEBLOADERSTRATEGY_RELEASE_LOG("scheduleLoad: URL will be scheduled with the NetworkProcess");
-    scheduleLoadFromNetworkProcess(resourceLoader, resourceLoader.request(), trackingParameters, shouldClearReferrerOnHTTPSToHTTPRedirect, maximumBufferingTime(resource));
+    const Seconds bufferingTime = ([&]{
+        if (resource && resource->type() == CachedResource::Type::RawResource && resourceLoader.request().timeoutInterval() > 0)
+            return Seconds(std::min(0.25, resourceLoader.request().timeoutInterval() / 2.0));
+        return maximumBufferingTime(resource);
+    })();
+    scheduleLoadFromNetworkProcess(resourceLoader, resourceLoader.request(), trackingParameters, shouldClearReferrerOnHTTPSToHTTPRedirect, bufferingTime);
 }
 
 bool WebLoaderStrategy::tryLoadingUsingURLSchemeHandler(ResourceLoader& resourceLoader, const WebResourceLoader::TrackingParameters& trackingParameters)
-- 
2.48.1

